src/lib/integration_test_local_engine/swarm_network.ml

```diff
@@ -0,0 +1,221 @@
+open Core
+open Async
+open Integration_test_lib
+
+(* exclude from bisect_ppx to avoid type error on GraphQL modules *)
+[@@@coverage exclude_file]
+
+module Node = struct
+  type t =
+    { swarm_name : string
+    ; service_id : string
+    ; graphql_enabled : bool
+    ; network_keypair : Network_keypair.t option
+    }
+
+  let id { service_id; _ } = service_id
+
+  let network_keypair { network_keypair; _ } = network_keypair
+
+  let get_container_cmd t =
+    Printf.sprintf "$(docker ps -f name=%s --quiet)" t.service_id
+
+  let run_in_postgresql_container _ _ = failwith "run_in_postgresql_container"
+
+  let get_logs_in_container _ = failwith "get_logs_in_container"
+
+  let run_in_container node cmd =
+    let base_docker_cmd = "docker exec" in
+    let docker_cmd =
+      Printf.sprintf "%s %s %s" base_docker_cmd (get_container_cmd node) cmd
+    in
+    let%bind.Deferred cwd = Unix.getcwd () in
+    Malleable_error.return (Util.run_cmd_exn cwd "sh" [ "-c"; docker_cmd ])
+
+  let start ~fresh_state node : unit Malleable_error.t =
```

Comment:
all the stuff with the /stop /start puppeteer scripts was kind of a hacky solution we had that mostly had specifically to do with problems with the GCP deployment.  mb @nholland94  can give more background about the gory details.  i think we should avoid this if we have time.  ideally, it would be best for the local engine to just figure out some docker command and make docker shut the service down and start it back up again, or something like that.

Reply:
While I do agree that a better solution would be preferable, I think the puppeteer scripts are actually extremely useful for getting this started. I'm not sure if we have plans to refactor the puppeteer scripts out of the cloud engine but being able to start all node containers via start and stop commands is really useful. I don't think there is a way to do this natively in docker-compose unless I missed something :/